spark/src/test/java/org/apache/iceberg/actions/TestRemoveFilesAction.java

```diff
@@ -0,0 +1,232 @@
+          Thread thread = new Thread(runnable);
+          thread.setName("drop-table-" + deleteThreadsIndex.getAndIncrement());
+          thread.setDaemon(true); // daemon threads will be terminated abruptly when the JVM exits
+          return thread;
+        }))
+        .deleteWith(s -> {
+          deleteThreads.add(Thread.currentThread().getName());
+          deletedFiles.add(s);
+        })
+        .execute();
+
+    // Verifies that the delete methods ran in the threads created by the provided ExecutorService ThreadFactory
+    Assert.assertEquals(deleteThreads,
+        Sets.newHashSet("drop-table-0", "drop-table-1", "drop-table-2", "drop-table-3"));
+
+    Assert.assertTrue("FILE_A should be deleted", deletedFiles.contains(FILE_A.path().toString()));
+    Assert.assertTrue("FILE_B should be deleted", deletedFiles.contains(FILE_B.path().toString()));
+
+    checkDropTableResults(4L, 6L, 4L, result);
+  }
+
+  @Test
+  public void testWithExpiringDanglingStageCommit() {
+    table.location();
+    // `A` commit
+    table.newAppend()
+        .appendFile(FILE_A)
+        .commit();
+
+    // `B` staged commit
+    table.newAppend()
+        .appendFile(FILE_B)
+        .stageOnly()
+        .commit();
+
+    // `C` commit
+    table.newAppend()
+        .appendFile(FILE_C)
+        .commit();
+
+
+    RemoveFiles.Result result = Actions.forTable(table).removeFilesAction()
+        .execute();
+
+    checkDropTableResults(3L, 3L, 3L, result);
+  }
+
+  @Test
+  public void testDropOnEmptyTable() {
+    RemoveFiles.Result result = Actions.forTable(table).removeFilesAction()
+        .execute();
+
+    checkDropTableResults(0, 0, 0, result);
+  }
+
+  @Test
+  public void testDropAction() {
+    table.newAppend()
+        .appendFile(FILE_A)
+        .commit();
+
+    table.newAppend()
+        .appendFile(FILE_B)
+        .commit();
+
+    BaseRemoveFilesSparkAction baseRemoveFilesSparkAction = Actions.forTable(table)
+        .removeFilesAction();
+    RemoveFiles.Result result = baseRemoveFilesSparkAction.execute();
+
+    checkDropTableResults(2, 2, 2, result);
+  }
+
+  @Test
+  public void testUseLocalIterator() {
+    table.newFastAppend()
+        .appendFile(FILE_A)
+        .commit();
+
+    table.newOverwrite()
+        .deleteFile(FILE_A)
+        .addFile(FILE_B)
+        .commit();
+
+    table.newFastAppend()
+        .appendFile(FILE_C)
+        .commit();
+
+    int jobsBefore = spark.sparkContext().dagScheduler().nextJobId().get();
+
+    RemoveFiles.Result results = Actions.forTable(table).removeFilesAction().option("stream-results", "true").execute();
+
+    int jobsAfter = spark.sparkContext().dagScheduler().nextJobId().get();
+    int totalJobsRun = jobsAfter - jobsBefore;
+
+    checkDropTableResults(3L, 4L, 3L, results);
+
+    Assert.assertTrue(
+        String.format("Expected total jobs to be equal to total number of shuffle partitions", SHUFFLE_PARTITIONS, totalJobsRun),
```

Comment:
@aokolnychyi @RussellSpitzer Since we dedup the values as part of the dataset, the number of elements in iterator will be equal to the number of shuffle partitions.